drv/lpc55-swd/src/main.rs

```diff
@@ -1058,21 +1297,626 @@ impl ServerImpl {
+        });
+
+        r
+    }
+
+    fn read_buf_from_addr(
+        &mut self,
+        addr: u32,
+        buf: &mut [u8],
+    ) -> Result<(), SpCtrlError> {
+        let start = addr;
+        let end = addr + buf.len() as u32;
+        self.start_read_transaction(start, ((end - start) as usize) / 4)
+            .map_err(|_| SpCtrlError::Fault)?;
+
+        let cnt = buf.len();
+        if cnt % 4 != 0 {
+            return Err(SpCtrlError::BadLen);
+        }
+
+        let mut i = 0usize;
+        for _ in 0..cnt / 4 {
+            match self.read_transaction_word() {
+                Ok(r) => {
+                    if let Some(w) = r {
+                        for b in w.to_le_bytes() {
+                            buf[i] = b;
+                            i += 1;
+                        }
+                    }
+                }
+                Err(_) => return Err(SpCtrlError::Fault),
+            }
+        }
 
         Ok(())
     }
+
+    fn do_setup_swd(&mut self) -> Result<(), SpCtrlError> {
+        ringbuf_entry!(Trace::DoSetup);
+
+        // TODO: clients interferring with each other and SP_RESET handling.
+        //
+        // A client can call into this SWD task using multiple API calls.
+        // While that session is active, inbetween calls, an SP_INTERRUPT
+        // can happen that will change the state of the SWD interface and the
+        // SP itself. If there were multiple external clients (hiffy vs dump),
+        // they could interfere with each other's work.
+        //
+        // The dump client is trying to be non-intrusive and will try to resume
+        // the SP after finishing its work, the last call being `resume` or, if
+        // that fails, `setup` followed by `resume`.
+        if !self.init {
+            ringbuf_entry!(Trace::PinSetupDefaults);
+            // This should be redundant since setup_pins needs to be
+            // called in order for SP_RESET interrupt to be plumbed.
+            // The only exception is if SP_RESET is driven as an output
+            // that that should be a non-interruptable transient state.
+            self.pin_setup();
+        }
+
+        if self.swd_dongle_detected() {
+            ringbuf_entry!(Trace::DongleDetected);
+            return Err(SpCtrlError::DongleDetected);
+        }
+
+        match self.swd_setup() {
+            Ok(_) => {
+                ringbuf_entry!(Trace::SwdSetupOk);
+                self.init = true;
+                Ok(())
+            }
+            Err(_) => {
+                ringbuf_entry!(Trace::SwdSetupFail);
+                Err(SpCtrlError::Fault)
+            }
+        }
+    }
+
+    fn do_halt(&mut self) -> Result<(), SpCtrlError> {
+        ringbuf_entry!(Trace::DoHalt);
+        self.dp_write_bitflags::<Dhcsr>(Dhcsr::halt())
+            .map_err(|_| SpCtrlError::Fault)?;
+        self.halt_wait(5000)
+    }
+
+    fn halt_wait(&mut self, timeout: u64) -> Result<(), SpCtrlError> {
```

Comment:
I think I confused myself because `halt_wait` means wait for the system to halt, not attempt to halt and then wait. Can you either add a comment to this effect or change the name?

Reply:
`wait_for_sp_halt` would be better.

Reply:
Changed